### PR TITLE
docs: Resolve Dagger version dynamically

### PR DESCRIPTION
**Pull request by gerhard**
This will fetch the version from https://dl.dagger.io/dagger/latest_version and template it in all doc pages on deploy.

Resolves https://github.com/dagger/dagger/issues/1854
Resolves https://github.com/dagger/dagger/issues/2205
Resolves https://github.com/dagger/dagger/issues/2351

The only thing which remains is to figure out how to template this version in the YAML snippets, e.g. https://github.com/dagger/dagger/pull/2369/files#diff-e309255ddd52101788f3de79596d0b93496c72d4b04059e44154f9e12b63e4d0R134

### CODE REVIEW

**Comment by aluzzardi** [comment]
I still don't understand the reasoning behind putting the dagger binary version inside the docs -- what problem are we solving?

- This and putting no version at all has the same effect
- ~Half our users are on macOS and installing through `brew` without using `DAGGER_VERSION`
- Docs **do** depend on the version of dagger, that is very true
  - For instance, the new output feature doc page can ONLY be followed by users with the **next** release which is not out yet: #2265 
  - The output feature will also affect other unrelated docs pages (e.g. if we display snippets of `dagger` and relying on outputs)
  - That's a problem and I agree with that. But the solution might be to add disclaimer banners (e.g. "requires dagger 0.2.9 or higher") or something. I don't think hardcoding the version helps with any of that

**Comment by slumbering** [comment]
> * This and putting no version at all has the same effect

@aluzzardi do you mean that : 
`curl -L https://dl.dagger.io/dagger/install.sh | sh` would do the same thing as `curl -L https://dl.dagger.io/dagger/install.sh | DAGGER_VERSION=0.2.9 sh`? No need to set DAGGER_VERSION to get the latest version ?
If yes, I agree that we definitely do not need to have this dynamic version mecanism.

**Comment by gerhard** [comment]
> I still don't understand the reasoning behind putting the dagger binary version inside the docs -- what problem are we solving?

We are hard-coding the version of the latest dagger CLI release in various places, as initially captured in #1854 (we have added a few more since then). We are doing this so that we have deterministic starting points for code examples, or linking to specific parts of our codebase.

> * This and putting no version at all has the same effect

No version means `main`, while a specific version means _this specific tag_

> * ~Half our users are on macOS and installing through `brew` without using `DAGGER_VERSION`

Using `curl` and a specific version means that `bin/dagger` will be that specific version. This can coexist with `brew`, and is a great solution to brew's inability to install a specific version.

>   * For instance, the new output feature doc page can ONLY be followed by users with the **next** release which is not out yet: [feat: Print action outputs to console #2265](https://github.com/dagger/dagger/pull/2265)

You are right, there are situations when we are publishing docs for an unreleased version of Dagger. In that case, we don't use the latest released version which gets dynamically resolved, but a specific version, e.g. `>= 0.2.9`.

>   * The output feature will also affect other unrelated docs pages (e.g. if we display snippets of `dagger` and relying on outputs)

Yes, that is definitely one of the follow-ups. In that case, I am wondering whether we want to keep an outdated version of dagger in the docs (since the output is linked to it), or we expect readers to notice the discrepancy and submit improvements to the examples, instead of the dagger versions (which is what has been happening in the last month).

**Comment by aluzzardi** [comment]
> No version means main, while a specific version means this specific tag.

I believe this only affects `1200-local-dev.md`, when doing `git checkout <VERSION>`, which could be replaced by a `git checkout $(dagger version --short)`.

No other documentation page would reference `main` if we were to loose `DAGGER_VERSION`.

`1200-local-dev.md` is even a special case since it mixes installing dagger and running an example at once, while it's not true for all the other documentation pages.

Were we following the same structure (`local-dev` is an example only), dynamically updating `git checkout v0.2.X` wouldn't help since it's entirely dependent on which version of dagger the user is using.

> Using curl and a specific version means that bin/dagger will be resolved to that specific version. This can coexist with brew, and is a great solution to brew's inability to install a specific version. It's also great being able to do this: https://github.com/thechangelog/changelog.com/blob/4291f328c53d89b63feb1c3affab580770927933/2022.fly/Makefile#L58-L64

Just on a single docs page. There's no relationship between `DAGGER_VERSION` and `1204-secrets.md`


Again, I don't understand the problem that's being solved. It would be helpful to see an example of a broken scenario that's fixed by DAGGER_VERSION.

- Making sure the user installs the latest version?
  - Removing DAGGER_VERSION does just that. With or without: same outcome
- Making sure the dagger binary is compatible with the docs page the user is viewing?
  - Still broken on all pages except `1200-local-dev.md`
  - Would also be broken for `1200-local-dev.md` if we were to have a separate `1234-install.md`
  - Dynamically checking out based on the user's dagger version would fit that
- Core Reference API?
  - The documentation is for a SPECIFIC version of Core
  - Pointing to any other tag than the one documented will be a mismatch
  - Whether it's `main` or the latest tag, it's the same sort of mismatch

**Comment by gerhard** [comment]
These are the problems that we have currently:
1. How do we link to latest released universe.dagger.io packages? https://github.com/dagger/dagger/pull/2404/files#diff-98af41b3bbc1c3486b8fea4c0e568d838bc562039bd94d56bc0515fec514a395 & https://github.com/dagger/dagger/pull/2404/files#diff-89d31acc694a4c8662162819393063442be9ae69922b13b0034968699e489640
1. How do we link to latest released dagger.io packages? https://github.com/dagger/dagger/pull/2404/files#diff-de1c0d68098ddfd5bd57ecdbb600e06fb554a41c10c6cd06f325e3336a947290
1. How do we install a specific dagger version? https://github.com/dagger/dagger/pull/2404/files#diff-8714bdb51cdd72074c3d20cacbd68b00500b3eeab44447f6418236496edf6bb0R49 & https://github.com/dagger/dagger/pull/2404/files#diff-e309255ddd52101788f3de79596d0b93496c72d4b04059e44154f9e12b63e4d0L70
1. How do we link to latest released Dagger CI workflow? https://github.com/dagger/dagger/pull/2404/files#diff-2e7838dd9e6f075d3028b6ed3dc1de37c8a95fa15c87086fcefa6c46b5102f4b

**Comment by gerhard** [comment]
Talking to @aluzzardi this is what we settled on:

1. Legit.
2. Lock to the version they were written for. When we update the core packages, we should update this version in the docs to reflect this. This may require a checklist.
3. Don't couple installing a specific version with installing the latest version. Replace specific dagger version in the CLI output with a placeholder, same as `GIT_SHA`
4. Legit.

Where we use `git checkout`, use `$(dagger version --short)` instead of a specific version. This will require implementing the `--short` flag.

In conclusion, continue as proposed with this.

**Comment by gerhard** [comment]
I am picking this one up again. Will ask for your review @slumbering as soon as everything is green. Would be great to get it merged today.

File: docs/getting-started/1201-ci-environment.md [hunk 1]
```diff
@@ -42,11 +45,13 @@ If you would like us to document CircleCI next, vote for it here: [dagger#1677](
 
 <TabItem value="gitlab">
 
-```yaml
```
[the remark this answers is not in the document]
Comment (gerhard) [reply docs/getting-started/1201-ci-environment.md:88]: I just got to the same issue now. Resolved all conflicts, including issues with my node install & node-fetch.

**Comment by gerhard** [comment]
This one has been going on for too long 🎡

We are planning to get it merged with @slumbering on July 25th. I'm going to solve the conflicts now and transition it from `draft` to `open` so that it's easier to push through come Monday.

**Review by slumbering — APPROVED** [review]
Let's merge it and fix the 1201-ci-environment in an other PR
